### PR TITLE
feat: Auto close chat if user logs out

### LINKED ISSUE #303: [FEATURE] Auto close chat when user logs out
### Description

my chat is not being closed when my buddy logs out. When my buddy log out, I expect the chat to be closed, and a confirmation dialog asking me if I would like to find a new buddy (just like the one I get when my buddy closes their chat) except if my buddy is logged in on multiple devices.

### Screenshots

_No response_

### Additional information

_No response_

### 👀 Have you checked if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read and understood the rules in the [Contributing Guidelines](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md)

### Are you willing to work on this issue ?

None

### PR DESCRIPTION
**Pull request by pateldivyesh1323**
# Fixes Issue #303 
- Auto close chats if one of the user logs out
- Note: There is currently no feature on the app to tell if a user is logged in on multiple devices, so if a user logs out from one device the chat is closed.

**My PR closes #303**

# ✔️ Check List (Check all the applicable boxes)

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.

### CODE REVIEW

**Comment by Dun-sin** [comment]
Good point, any suggestions on how to keep track of different devices users are logged into?

**Comment by pateldivyesh1323** [comment]
@Dun-sin I think so, we can maintain a session count in userschema, if a user login and logout we can increment and decrement it accordingly. So, if session count is 0 we will come to know that user is not logged in from any other device.

**Comment by Dun-sin** [comment]
> @Dun-sin I think so, we can maintain a session count in userschema, if a user login and logout we can increment and decrement it accordingly. So, if session count is 0 we will come to know that user is not logged in from any other device.

Good logic but there's a problem with that, what if the user isn't active on one but is logged in, we'll be keeping someone hanging

**Comment by pateldivyesh1323** [comment]
@Dun-sin then we should go with my current implementation, if user logs out even from one device, the chat is closed, why a user will log out if he wants to chat.

**Comment by Dun-sin** [comment]
> @Dun-sin then we should go with my current implementation, if user logs out even from one device, the chat is closed, why a user will log out if he wants to chat.

Yes looks like we have to maintain that logic

**Comment by Dun-sin** [comment]
this doesn't work, how did you test that it works, a description and a video would be good

**Comment by pateldivyesh1323** [comment]
Strange, this works perfectly fine locally.

[video-convert-1696178479540.webm](https://github.com/Dun-sin/Whisper/assets/109150688/7c14f783-5c58-412b-ad87-b930c33ff57d)

**Comment by Dun-sin** [comment]
> Strange, this works perfectly fine locally.
> video-convert-1696178479540.webm

the “it works on my pc” meme is acting out right now, how ironic, lol
here's a video of it working deployed you can also find the URL in this PR, vercel automatically deploys it


https://github.com/Dun-sin/Whisper/assets/78784850/265b1e54-9bd5-497d-bb4b-908c45ed001d

**Comment by pateldivyesh1323** [comment]
@Dun-sin 😅Sorry for that, I did not do anything intentionally, and now I don't see a way to debug it.

File: server/sockets/logout.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
-const { NEW_EVENT_LOGOUT, NEW_EVENT_INACTIVE } = require("../../constants.json");
-const { getActiveUser, delWaitingUser } = require("../utils/lib");
+const { NEW_EVENT_LOGOUT, NEW_EVENT_INACTIVE, NEW_EVENT_CLOSE } = require("../../constants.json");
+const { getActiveUser, delWaitingUser, closeChat } = require("../utils/lib");
```
Comment (Dun-sin) [inline server/sockets/logout.js:2]: why did you import closeChat here

Comment (pateldivyesh1323) [reply server/sockets/logout.js:2]: Because, it was used on line 18 and not imported.

Comment (Dun-sin) [reply server/sockets/logout.js:2]: if it was used and not imported we would have got an error a long time ago, remove the change for that line, I suspect it to be the problem also it has nothing to do with your PR

Comment (Dun-sin) [reply server/sockets/logout.js:2]: hmmmm, tough spot

**Comment by Dun-sin** [comment]
@pateldivyesh1323 I can assign this to someone else then??

**Comment by pateldivyesh1323** [comment]
@Dun-sin yes please.

**Pull request closed without merging**